pkg/kubelet/cm/device_plugin_handler.go

```diff
@@ -0,0 +1,257 @@
+		var capacity = v1.ResourceList{}
+		kept := append(updated, added...)
+		if _, ok := handler.allDevices[resourceName]; !ok {
+			handler.allDevices[resourceName] = sets.NewString()
+		}
+		// For now, DevicePluginHandler only keeps track of healthy devices.
+		// We can revisit this later when the need comes to track unhealthy devices here.
+		for _, dev := range kept {
+			if dev.Health == pluginapi.Healthy {
+				handler.allDevices[resourceName].Insert(dev.ID)
+			} else {
+				handler.allDevices[resourceName].Delete(dev.ID)
+			}
+		}
+		for _, dev := range deleted {
+			handler.allDevices[resourceName].Delete(dev.ID)
+		}
+		capacity[v1.ResourceName(resourceName)] = *resource.NewQuantity(int64(handler.allDevices[resourceName].Len()), resource.DecimalSI)
+		updateCapacityFunc(capacity)
+	}
+
+	mgr, err := deviceplugin.NewManagerImpl(pluginapi.KubeletSocket, deviceManagerMonitorCallback)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to initialize device plugin manager: %+v", err)
+	}
+
+	handler.devicePluginManager = mgr
+	return handler, nil
+}
+
+func (h *DevicePluginHandlerImpl) Start() error {
+	return h.devicePluginManager.Start()
+}
+
+func (h *DevicePluginHandlerImpl) Devices() map[string][]*pluginapi.Device {
+	return h.devicePluginManager.Devices()
+}
+
+func (h *DevicePluginHandlerImpl) Allocate(pod *v1.Pod, container *v1.Container, activePods []*v1.Pod) ([]*pluginapi.AllocateResponse, error) {
+	glog.V(3).Infof("Allocate called")
+	var ret []*pluginapi.AllocateResponse
+	h.Lock()
+	defer h.Unlock()
+	h.updateAllocatedDevices(activePods)
+	for k, v := range container.Resources.Limits {
+		key := string(k)
+		needed := int(v.Value())
+		glog.V(3).Infof("needs %d %s", needed, key)
+		if !deviceplugin.IsDeviceName(k) || needed == 0 {
+			continue
+		}
+		// Gets list of devices that have already been allocated.
+		// This can happen if a container restarts for example.
+		if h.allocatedDevices[key] == nil {
+			h.allocatedDevices[key] = newPodDevices()
+		}
+		devices := h.allocatedDevices[key].getDevices(string(pod.UID), container.Name)
+		if devices != nil {
+			glog.V(3).Infof("Found pre-allocated devices for container %q in Pod %q: %v", container.Name, pod.UID, devices.List())
+			needed = needed - devices.Len()
+		}
+		// Get Devices in use.
+		devicesInUse := h.allocatedDevices[key].devices()
+		glog.V(3).Infof("all devices: %v", h.allDevices[key].List())
+		glog.V(3).Infof("devices in use: %v", devicesInUse.List())
+		// Get a list of available devices.
+		available := h.allDevices[key].Difference(devicesInUse)
+		glog.V(3).Infof("devices available: %v", available.List())
+		if int(available.Len()) < needed {
+			return nil, fmt.Errorf("requested number of devices unavailable. Requested: %d, Available: %d", needed, available.Len())
+		}
+		allocated := available.UnsortedList()[:needed]
+		resp, err := h.devicePluginManager.Allocate(key, append(devices.UnsortedList(), allocated...))
```

Comment:
@RenaudWasTaken @jiayingz  one suggestion, I think we should invoke `h.devicePluginManager.Allocate` outside of mutex lock because plugin manager is already using locks. Mutex of device plugin handler should be used only to access handler's internal data structures like `h.allocatedDevices` for example. This looks to me performance efficient than existing implementation. Thoughts?

Reply:
+1. A remote RPC with a lock is probably not a good idea.

Reply:
Done.